flink-core/src/main/java/org/apache/flink/api/connector/source/lib/util/RateLimitedSourceReader.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.api.connector.source.lib.util;
+
+import org.apache.flink.annotation.Experimental;
+import org.apache.flink.api.connector.source.ReaderOutput;
+import org.apache.flink.api.connector.source.SourceReader;
+import org.apache.flink.api.connector.source.SourceSplit;
+import org.apache.flink.core.io.InputStatus;
+
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/** Wraps the actual {@link SourceReader} and rate limits its data emission. */
+@Experimental
+public class RateLimitedSourceReader<E, SplitT extends SourceSplit>
+        implements SourceReader<E, SplitT> {
+
+    private final SourceReader<E, SplitT> sourceReader;
+    private final RateLimiter rateLimiter;
+
+    /**
+     * Instantiates a new rate-limited source reader.
+     *
+     * @param sourceReader The actual source reader.
+     * @param rateLimiter The rate limiter.
+     */
+    public RateLimitedSourceReader(SourceReader<E, SplitT> sourceReader, RateLimiter rateLimiter) {
+        checkNotNull(sourceReader);
+        checkNotNull(rateLimiter);
+        this.sourceReader = sourceReader;
+        this.rateLimiter = rateLimiter;
+    }
+
+    // ------------------------------------------------------------------------
+
+    @Override
+    public void start() {
+        sourceReader.start();
+    }
+
+    @Override
+    public InputStatus pollNext(ReaderOutput<E> output) throws Exception {
+        rateLimiter.acquire();
```

Comment:
is the reader thread allowed to block? The javadocs state this method must be non-blocking.

How would a non-blocking source communicate that data is available at a later time + that it has become available?

Reply:
I am afraid that if we want to 100% adhere to the contract of the "proper" source for this "mock" source, we'd have to drop rate limiting. Having a non-blocking trotting built-in into the Source API looks like a major effort that would require a separate design discussion and a FLIP. I know there where some discussions about this in the mailing list, but the did not seem to have materialized into anything yet:
https://lists.apache.org/thread/93hwlq8lzwx2wvlh9wvsv23hthcx6y3n
My plan was to go with the basic functionality and converge to the "final" solution once it is there from the Source API side.
Dropping the rate limiting would be unfortunate. In my eyes it makes the `DataGeneratorSource` half as useful.

Reply:
Or we could just implement the rate-limiting in a non-blocking fashion.

Add a (single-threaded) Executor to the rate limiter that does the acquire call against guava, and return a future to the source reader that is completed once the permits were acquired.

Reply:
I can try to sketch this out tomorrow.

Reply:
See https://github.com/zentol/flink/commit/0c2b75317d3be15ab502ae39b600f494422e3474

Reply:
Thanks for the sketch. A couple of thoughts:
- Can't we use [tryAcquire()](https://guava.dev/releases/19.0/api/docs/com/google/common/util/concurrent/RateLimiter.html#tryAcquire()) for such implementation directly?
- Let's say we set the count to Long.MAX_VALUE. The call to `sourceReader.pollNext(output)` is in this case unrestricted and happens in a busy loop. The underlying sourceReader is going to keep internally calling `output.collect(iterator.next())` without any limitations. Won't it blow up the underlying queues/futures?
- If the capacity is not an issue, if I am not mistaken, this will still cause the number of cores equal to the configured parallelism to go into a 100% load state even when configured to run with, say, 10 records per second. That might be surprising for users who just want to run a demo/quick test locally. Do I get it wrong?

Reply:
> Can't we use tryAcquire() for such implementation directly?

No, that would imply hot-looping if there if we have currently hit the rate limit.

> The call to sourceReader.pollNext(output) in the sketch is unrestricted

That's not quite true. The contract of the `SourceReader`, as I understand it, is that if you return `NOTHING_AVAILABLE` then `isAvailable()` is called, and `pollNext()` is only called again once the returned future is complete.
By never returning `MORE_AVAILABLE` `pollNext()` should never be called multiple times in a row.

> The underlying sourceReader is going to keep internally calling output.collect(iterator.next()) without any limitations

This is also an issue in the current PR. SourceReaders are also encouraged, per the javadocs, to not emit multiple records within a single `pollNext()` call.

>  Do I get it wrong?

Yes, I think so. There's no busy waiting because we either a) wait for the future to complete or b) poll values.

Reply:
OK, nice, that makes sense, I misinterpreted the interplay between the isAvailable() and pollNext().

Reply:
Thanks for the proposed adjustments, I've integrated them into this branch.